pyo3-build-config/src/impl_.rs

```diff
@@ -172,6 +172,11 @@ impl InterpreterConfig {
                     See https://foss.heptapod.net/pypy/pypy/-/issues/3397 for more information."
                 );
             }
+        } else if self.implementation.is_graalpy() {
+            println!("cargo:rustc-cfg=GraalPy");
+            if self.abi3 {
+                warn!("GraalPy does not support abi3 so the build artifacts will be version-specific.");
```

Comment:
AFAIK ABI3 exposes both `ob_refcnt` and `ob_type`. Since we have a completely different object layout and GC, it will be very difficult for us to support this in any way. Our current `PyObject*` that we are handing out to C extensions for managed Python objects are not even valid pointers, they are opaque handles that if resolved point into kernel space. That's why I also changed the pyo3-ffi code to not access ob_refcnt and some other fields directly, but always go through a function call on GraalPy

Reply:
Understood, so we are waiting for a future abi4 or so for GraalPy then 👍 

Out of curiosity do you have support for HPy?

Reply:
> Out of curiosity do you have support for HPy?

Yes, we contribute to HPy and have complete support. We discussed porting PyO3 to use HPy and proposing those changes here in the GraalPy team, but it's a bigger undertaking and we wanted to get some support earlier

Reply:
> Our current PyObject* that we are handing out to C extensions for managed Python objects are not even valid pointers, they are opaque handles that if resolved point into kernel space. That's why I also changed the pyo3-ffi code to not access ob_refcnt and some other fields directly, but always go through a function call on GraalPy

From just reading this, I wonder if this brings us into insta-UB territory as our `&'py PyAny` would then be an invalid reference?

Reply:
@adamreichold do you mean the pointers to kernel space? i don't see how that can be undefined behaviour, since the compiler cannot speculate on whether or not I will have kernel privileges when running any particular piece of code, so it's not like the compiler can assume a pointer to kernel space can fault immediately. Furthermore, why would the compiler even generate code to inspect the pointer or make assumptions about where it points? it receives it via ffi from an external source. why would it insert code to inspect each pointer for validity before storing it?

Reply:
maybe I should not have written "not even valid pointers" above - they are of course valid pointers. it just so happens that if you try to deference them in a user process, you will get a segfault. which is why we never deference them when compiling for GraalPy

Reply:
thinking about it some more, we could also give you some pointers into a page in the current process to represent our PyObject*, and we simply mprotect them to deny all access after we handed you those pointers. it wouldn't be any different afaics and the rust compiler couldn't know this either.

Reply:
I thin also that #3382 will help reassure the problem here because then we won't actually convert the pointers to references, we will just be carrying them around inside a wrapper type. Given that we're not planning to release before then, if this is passing CI I'm sorta ok with not worrying about this particular concern.

Reply:
> I thin also that https://github.com/PyO3/pyo3/issues/3382 will help reassure the problem here because then we won't actually convert the pointers to references,

I thought so too at first, but as long as `Py2::into_gil_ref` and friends exist we still have that problem. And even when those are gone because everything uses `Py2`, I think we should add additional compile-time/by-design guard rails to avoid turning these pointers into even short-lived references unintentionally if we really want to be "handle safe".

> maybe I should not have written "not even valid pointers" above - they are of course valid pointers. it just so happens that if you try to deference them in a user process, you will get a segfault. which is why we never deference them when compiling for GraalPy

The Rust compilers has additional requirements for references and violating them is instant undefined behaviour, i.e. undefined behaviour as soon as the reference is constructed, independently of whether it is ever used.

This includes that references must non-null and properly aligned for the type and cannot dangle. IIRC, this generally related to optimizations and specifically to various LLVM attributes like "dereferencable" attached when these values are passed into functions.

To be honest, I am not sure whether the kernel-space or mprotect tricks run afoul of that because the OS-provided address space shenanigans are outside of the scope of Rust's abstract machine (don't ask about `mmap` either).

Short term, I would be glad if you could get some expert advice over at [Rust's UCG](https://github.com/rust-lang/unsafe-code-guidelines) on exactly this problem.

Long term and also for HPy support, I think it would be best if we avoided using `*mut PyObject` as a type internally and would really use an opaque handle type (even if is a newtype wrapper of `*mut PyObject` on CPython) that is impossible to turn into a reference by accident even in unsafe code.

Reply:
(As an example of what can go wrong: Since the compiler can assume alignedness, it can clobber the lower bits with random values as long as it zeros them out again before passing the reference elsewhere.)

Reply:
Fair point, I'll find a moment to reach the UCG before we merge. Also even if `into_gil_ref` continues to exist for a few versions we can gate it on a feature which disallows `GraalPy` use.

I suppose we should also document on `Py::into_ptr` and `Py::as_ptr` methods (and same for `Py2`) that depending on the runtime this pointer may be just an opaque handle and is never safely dereferenceable? Probably best to do that in this PR.

Reply:
One solution that would help a lot here would be to define `ffi::PyObject` as an `opaque_struct!` on `GraalPy`, if these pointers are never derefenceable? Then the type system basically bans us from doing any reads and the compiler treats it as a zero-sized typed. @timfel does that make sense in your eyes?

Reply:
> (As an example of what can go wrong: Since the compiler can assume alignedness, it can clobber the lower bits with random values as long as it zeros them out again before passing the reference elsewhere.)

Yes, we ensure that the pointers we hand out are aligned, and we don't use the upper 16 bits either. So zero-ing out should not be an issue.

> To be honest, I am not sure whether the kernel-space or mprotect tricks run afoul of that because the OS-provided address space shenanigans are outside of the scope of Rust's abstract machine (don't ask about mmap either).

Mh, interesting. I'm completely new to Rust, but I wonder what that means for the story of Rust interfacing with managed languages and JIT compilers in general, which like to do such stuff to intercept accesses to code during concurrent deoptimization for example. I guess Rust code just must not store references to memory that can be subject to this, but that sounds to me like any Rust code must know an awful lot about internals of those runtimes it interfaces with. This may also be interesting down the line with CPython given the progress on evolving the C API to use completely opaque pointers which will then allow pointer tagging similar to how Ruby already does it.

> One solution that would help a lot here would be to define ffi::PyObject as an opaque_struct! on GraalPy, if these pointers are never derefenceable? Then the type system basically bans us from doing any reads and the compiler treats it as a zero-sized typed. @timfel does that make sense in your eyes?

I'm not entirely sure what that means, but "opaque" sounds right here :)

Reply:
@timfel our `opaque_struct!` macro is the recommended best practice for denoting extern types where we don't want to specify any layout. See for example

https://github.com/PyO3/pyo3/blob/dc251d0c9fe66f977f60ecd0fa59c3e135c26354/pyo3-ffi/src/object.rs#L6-L7

So what I'm suggesting is that we add in this PR

```rust
#[cfg(GraalPy)]
opaque_struct!(PyObject)
```

>  I wonder what that means for the story of Rust interfacing with managed languages and JIT compilers in general

This is reaching beyond my knowledge too, but from the best of my understanding the compiler won't make assumptions beyond what the definition of the type we give it implies. There's a fairly long discussion at  https://doc.rust-lang.org/nomicon/ffi.html#representing-opaque-structs (which is where our `opaque_struct!` helper came from).

Reply:
So, two things:
1) re `opaque_struct!(PyObject)` - I'm working on this, but it spreads throughout everything and I'm trying to get the tests running with it
2) For the next release of GraalPy, we are changing it so we actually allocate an arena of size PyVarObject and then hand out pointers to those. So the pointers would be real pointers, to dereferencable memory, with enough size. The type pointers will be correct, but ob_refcnt may be random and writing to the memory will not be reflected in the runtime, so on `#[cfg(GraalPy)]` we will continue to disable all direct accesses into the structs. Would that be ok w.r.t. UD @adamreichold ?

Reply:
> Would that be ok w.r.t. UD @adamreichold ?

From my point of view, yes. But I admittedly do not consider myself an expert on the intricacies of the Rust abstract machine, especially since those are still very much in flux. Hence, I would recommend that you do not take on this work based only on my concerns but get an expert opinion over at the Unsafe Code Working Group because they might tell you with some authority that you are currently is not a problem. With me, it's mostly fear, uncertainty and doubt...

Reply:
WRT 1. and the fallout of that change - yes I imagined it would force us to disable all code which ever dereferences the pointer, which would be quite a bit of code. At least it gives us a lot more confidence the pointer is never dereferenced 😄

Reply:
I imagine the best way to do that would be to make all those structs opaque too (maybe that's what you're doing)?

Reply:
> I imagine the best way to do that would be to make all those structs opaque too (maybe that's what you're doing)?

I'm trying, but for example `PyTypeObject` and `PyModuleDef` are not opaque for us. The latter cannot be opaque, because that's how an extension declares the module structure for the interpreter.

Reply:
I opened an issue for the Unsafe Code Working Group to clarify: https://github.com/rust-lang/unsafe-code-guidelines/issues/473

The Rust book seems to confirm that the compiler can insert spurious reads:

From https://doc.rust-lang.org/stable/std/cell/struct.UnsafeCell.html
> However, whenever a &UnsafeCell<T> is constructed or dereferenced, it must still point to live memory and the compiler is allowed to insert spurious reads if it can prove that this memory has not yet been deallocated.

Your `opaque_struct!` macro doesn't actually fix that, since according to https://doc.rust-lang.org/nomicon/ffi.html#representing-opaque-structs, while these structures are unknown, they are still assumed to point to readable memory on construction of the `UnsafeCell`.

I will wait and see what responses come on the issue, but the way I understand it you are already relying on a current implementation detail of CPython here when it comes to the limited API to not get UB. abi3 wheels you are building may not actually be forward compatible if (for example) CPython started to do things like tagged integers - in the limited API case they can do that and wheels using plain C are forward compatible, because no such UB exists, but they would not be for Rust extensions compiled with the current PyO3

Reply:
Where does that leave the `opaque_struct` change? Are you now feeling that's not worth pursuing, pending the outcome of the question to the working group?

Reply:
Ok, so we a clarification from upstream, there cannot be any `&PyAny` (whether short-lived or GIL-bound) where the underlying PyObject is located protected memory, i.e. reading from the pointer would trap, as the compiler is allowed to insert spurious reads.

So to be sound, shipping this needs to wait until there are no more bare references into the Python heap, i.e. the transition to `Py2<'py, T>` is done. Or GraalPy must back the references by dummy objects?

Reply:
I will spend some more time on the opaque structs, but it's turning into a huge change, because of the structures that cannot be opaque. As I said, we'll be backing our handles with readable dummy memory for our release in April, at which point all of the UB would go away.

Reply:
@timfel if it's not looking easy I suggest you wait for our change to #3382 which I hope to finish by the end of the year. We can ensure graalpy is only usable on that API, which won't ever create these references

Reply:
One other idea, but not sure if you would be happy to accept that: specifically quoting from https://github.com/rust-lang/unsafe-code-guidelines/issues/473#issuecomment-1792842516

> EXPLICITLY NON-NORMATIVE: This is in a class of "probably okay, but don't count on it working forever" UB. This is because while it's UB in Rust semantics, it's not UB in the LLVM IR as we currently emit it. This is because the LLVM attribute dereferencable(N) lasts until function exit, and LLVM does not provide a dereferencable_on_entry(N) attribute yet. But the intention is that if/when it gets one to use it for &UnsafeCell.

So currently this is *not* UB, provided there really are not any reads to the struct fields, because then the compiler could re-order them and cause UB. Would it be ok to insert e.g. at https://github.com/PyO3/pyo3/pull/3247/files#diff-5f4a9955ffbc1038032f0905b2fb70e7dbd991e489532e4529fe14a576aa0a25R677 something to bail when the rustc version is newer than the current latest (which we know is still using LLVM IR that doesn't insert these spurious reads?). We can then drop that condition again in April once we have a GraalPy release that backs objects with proper memory.

Reply:
The downside of that is 6 weeks after we release nobody will be able to compile their extension with latest stable Rust. That will cause a headache for distribution.

Reply:
> The downside of that is 6 weeks after we release nobody will be able to compile their extension with latest stable Rust. That will cause a headache for distribution.

You're right, ignore that suggestion, I didn't think that through